src/cpp/pipelinereply.cpp

```diff
@@ -0,0 +1,109 @@
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "pipelinereply.h"
+#include "srexception.h"
+
+using namespace SmartRedis;
+
+// Move constructor with QueuedReplies
+PipelineReply::PipelineReply(QueuedReplies&& reply)
+{
+    _add_queuedreplies(std::forward<QueuedReplies>(reply));
+}
+
+// Move assignment with QueuedReplies
+PipelineReply& PipelineReply::operator=(QueuedReplies&& reply)
+{
+    _queued_replies.clear();
+    _all_replies.clear();
+    _add_queuedreplies(std::forward<QueuedReplies>(reply));
+    return *this;
+}
+
+// Add QueuedReplies content to Pipeline ojbect via move semantics
+void PipelineReply::operator+=(QueuedReplies&& reply)
+{
+    _add_queuedreplies(std::forward<QueuedReplies>(reply));
+}
+
+// Return a shallow copy of an entry in the PipelineReply
+CommandReply PipelineReply::operator[](size_t index)
+{
+    if (index > _all_replies.size()) {
+        throw SRInternalException("An attempt was made to access index " +
+                                  std::to_string(index) +
+                                  " of the PipelineReply, which is beyond the "\
+                                  " PipelineReply length of " +
+                                  std::to_string(_all_replies.size()));
+    }
+
+    return CommandReply::shallow_clone(_all_replies[index]);
+}
+
+// Get the number of CommandReply in the PipelineReply
+size_t PipelineReply::size()
+{
+    return _all_replies.size();
+}
+
+// Check if any CommandReply has error
+bool PipelineReply::has_error()
+{
+    for (size_t i = 0; i < _all_replies.size(); i++) {
+        if (CommandReply::shallow_clone(_all_replies[i]).has_error() > 0)
```

Comment:
Seems wasteful to make a clone when all we need to do is read a field

Reply:
The ``_all_replies`` variable stores ``redisReply*`` types.  To re-use the recursive error checking code in CommandReply, I opted for the shallow copy (which is basically one pointer set to NULL and one pointer copy) over re-rewriting error checking code.  I'm open to rewriting the error checking code if we think it's a big perf hit.